source/http_server.c

```diff
@@ -0,0 +1,228 @@
+    PyObject *py_socket_options = NULL;
+    PyObject *tls_conn_options_capsule = NULL;
+
+    if (!PyArg_ParseTuple(
+            args,
+            "OOOs#HOO",
+            &bootstrap_capsule,
+            &on_incoming_connection,
+            &on_destroy_complete,
+            &host_name,
+            &host_name_len,
+            &port_number,
+            &py_socket_options,
+            &tls_conn_options_capsule)) {
+        PyErr_SetNone(PyExc_ValueError);
+        goto error;
+    }
+
+    if (host_name_len >= (signed int)AWS_ADDRESS_MAX_LEN || host_name_len == 0) {
+        PyErr_SetString(PyExc_ValueError, "host_name is not valid");
+        goto error;
+    }
+
+    if (py_socket_options == Py_None) {
+        PyErr_SetString(PyExc_ValueError, "socket_options is a required argument");
+        goto error;
+    }
+
+    if (!PyCallable_Check(on_incoming_connection)) {
+        PyErr_SetString(PyExc_ValueError, "on_incoming_connection callback is required");
+        goto error;
+    }
+
+    if (!PyCallable_Check(on_destroy_complete)) {
+        PyErr_SetString(PyExc_TypeError, "on_destroy_complete is invalid");
+        goto error;
+    }
+
+    struct aws_server_bootstrap *bootstrap = PyCapsule_GetPointer(bootstrap_capsule, s_capsule_name_server_bootstrap);
+    if (!bootstrap) {
+        PyErr_SetString(PyExc_ValueError, "the bootstrap capsule has an invalid pointer");
+        goto error;
+    }
+
+    struct aws_tls_connection_options *connection_options = NULL;
+
+    if (tls_conn_options_capsule != Py_None) {
+        connection_options = PyCapsule_GetPointer(tls_conn_options_capsule, s_capsule_name_tls_conn_options);
+    }
+
+    py_server = aws_mem_calloc(allocator, 1, sizeof(struct py_http_server));
+    if (!py_server) {
+        PyErr_SetAwsLastError();
+        goto error;
+    }
+
+    struct aws_socket_options socket_options;
+    AWS_ZERO_STRUCT(socket_options);
+    if (!aws_socket_options_init_from_py(&socket_options, py_socket_options)) {
+        goto error;
+    }
+
+    Py_INCREF(on_incoming_connection);
+    py_server->on_incoming_connection = on_incoming_connection;
+
+    Py_INCREF(on_destroy_complete);
+    py_server->on_destroy_complete = on_destroy_complete;
+    
+    py_server->allocator = allocator;
+
+    struct aws_http_server_options options;
+    AWS_ZERO_STRUCT(options);
+    options.self_size = sizeof(options);
+    options.bootstrap = bootstrap;
+    options.tls_options = connection_options;
+    options.allocator = allocator;
+    options.server_user_data = py_server;
+    struct aws_socket_endpoint endpoint;
+    AWS_ZERO_STRUCT(endpoint);
+
+    snprintf(endpoint.address, host_name_len, "%s", host_name);
+    endpoint.port = port_number;
+    options.socket_options = &socket_options;
+    options.endpoint = &endpoint;
+    options.on_incoming_connection = s_on_incoming_connection;
+    options.on_destroy_complete = s_on_destroy_complete;
+    py_server->server = aws_http_server_new(&options);
```

Comment:
not checking results of aws_http_server_new()

Reply:
I guess if(py_server->server) check the result of that new function